lightning-net/src/lib.rs

```diff
@@ -0,0 +1,1146 @@
+                // This should always succeed since send_data just pushes data
+                // into the writer_cmd channel and the writer_cmd channel always
+                // starts out completely empty. If pushing the initial 10s of
+                // bytes into the writer_cmd channel fails, something is very
+                // wrong; probably a programmer error.
+                if bytes_pushed != initial_data.len() {
+                    panic!("The initial write should always succeed");
+                }
+            }),
+    }
+    .map(|()| {
+        // PeerManager accepted the connection; kick off processing by spawning
+        // the Reader / Writer threads.
+        thread::spawn(move || reader.run());
+        thread::spawn(move || writer.run());
+    })
+    .map_err(|e| {
+        // PeerManager rejected this connection; disconnect the TcpStream and
+        // don't even start the Reader and Writer.
+        descriptor.disconnect_socket();
+        // In line with the requirements of new_inbound_connection() and
+        // new_outbound_connection(), we do NOT call socket_disconnected() here.
+        e
+    })
+}
+
+// NOTE: It would be nice to have a `impl From<SocketAddr> for NetAddress` in
+//       the `lightning` crate
+fn to_net(socket_addr: SocketAddr) -> NetAddress {
+    match socket_addr.ip() {
+        IpAddr::V4(ip) => NetAddress::IPv4 {
+            addr: ip.octets(),
+            port: socket_addr.port(),
+        },
+        IpAddr::V6(ip) => NetAddress::IPv6 {
+            addr: ip.octets(),
+            port: socket_addr.port(),
+        },
+    }
+}
+
+/// Commands that can be sent to the Reader.
+enum ReaderCommand {
+    ResumeRead,
+    PauseRead,
+    Shutdown,
+}
+
+/// Commands that can be sent to the Writer.
+enum WriterCommand {
+    WriteData(Vec<u8>),
+    Shutdown,
+}
+
+/// Initializes the crossbeam channels for sending `ReaderCommand`s and
+/// `WriterCommand`s.
+///
+/// The `ReaderCommand` channel is unbounded, and can be used to tell the
+/// `Reader` to resume reads, pause reads, or shut down.
+///
+/// The `WriterCommand` channel has a capacity of 2, and can be used to tell the
+/// `Writer` to write a Vec<u8> of data, or shut down.
+///
+/// - The WriterCommand channel is size 2 so as to have 1 dedicated space for
+///   each type of WriterCommand: WriteData and Shutdown. A Shutdown command can
+///   be pushed into the channel at any time, but to ensure that there is always
+///   space for it, send_data() will only ever push a WriteData command into the
+///   channel after it first detects that the channel is completely empty.
+/// - The reason we can't have a second dedicated channel for sending Shutdown
+///   commands to the Writer is that, being in a synchronous context, the Writer
+///   can only block on one channel at once, which means that in the case of two
+///   channels it could wait for WriteData commands or Shutdown commands, but
+///   not both.
```

Comment:
Gotcha - I'll just delete the comment then, since the main problem is that we can't select across both `channel_rx.recv()` and `TcpStream::read()`/`write()`.